engine/src/main/java/org/hibernate/validator/internal/util/ReflectionHelper.java

```diff
@@ -499,4 +498,85 @@ public static Type boxedType(Type type) {
 
 		return wrapperType;
 	}
+
+	/**
+	 * Builds an {@link Iterator} for a given array. It is (un)necessarily ugly because we have to deal with array of primitives.
+	 *
+	 * @param object a given array
+	 * @return an {@code Iterator} iterating over the array
+	 */
+	private static Iterator<?> buildIteratorFromArray(Object object) {
+		Iterator<?> iterator;
+		if ( Object.class.isAssignableFrom( object.getClass().getComponentType() ) ) {
+			iterator = Arrays.asList( (Object[]) object ).iterator();
+		}
+		else if ( object.getClass() == boolean[].class ) {
```

Comment:
Ok, if we need to do all that, how about avoiding the conversion to list then alltogether and have explicite iterator types:

    private static class BooleanArrayIterator implements Iterator<Boolean> {

        private boolean[] values;
        private int current = 0;

        public BooleanArrayIterator(boolean[] values) {
            this.values = values;
        }

        @Override
        public boolean hasNext() {
            return current < values.length;
        }

        @Override
        public Boolean next() {
            boolean result = values[current];
            current++;
            return result;
        }
    }